primitive array resultType throws exception

This issue has been reported previously at https://github.com/mybatis/mybatis-3/issues/259 and fixed.
But I think it is still broken.
mybatis version= 3.3.0

I have
&nbsp;&nbsp;&nbsp;int[] getValues(...
mapped to
&nbsp;&nbsp;&nbsp;&lt;select id="getValues" resultType="int"&gt;

I get this exception
Caused by: java.lang.ClassCastException: [I cannot be cast to [Ljava.lang.Object;
&nbsp;&nbsp;&nbsp;at org.apache.ibatis.binding.MapperMethod.convertToArray(MapperMethod.java:144)

At line 144 of MapperMethod.java you are trying to cast int[] to Object[] since E resolves to Object
E[] array = (E[]) Array.newInstance(method.getReturnType().getComponentType(), list.size());

--- a/src/main/java/org/apache/ibatis/binding/MapperMethod.java
+++ b/src/main/java/org/apache/ibatis/binding/MapperMethod.java
@@ -167,10 +167,22 @@ public class MapperMethod {
   }
 
   @SuppressWarnings("unchecked")
-  private <E> E[] convertToArray(List<E> list) {
-    E[] array = (E[]) Array.newInstance(method.getReturnType().getComponentType(), list.size());
-    array = list.toArray(array);
-    return array;
+  private <E> Object convertToArray(List<E> list) {
+
+    Class<?> arrayComponentType = method.getReturnType().getComponentType();
+
+    Object array = Array.newInstance(arrayComponentType, list.size());
+    Object value;
+    if (arrayComponentType.isPrimitive()) {
+      for (int i = 0; i < list.size(); i++) {
+        Array.set(array, i, list.get(i));
+      }
+      value = array;
+    } else {
+      value = list.toArray((Object[])array);
+    }
+
+    return value;
   }
 
   private <K, V> Map<K, V> executeForMap(SqlSession sqlSession, Object[] args) {

--- a/src/test/java/org/apache/ibatis/submitted/array_result_type/ArrayResultTypeTest.java
+++ b/src/test/java/org/apache/ibatis/submitted/array_result_type/ArrayResultTypeTest.java
@@ -88,12 +88,10 @@ public class ArrayResultTypeTest {
     }
   }
 
-  @Test(expected = ClassCastException.class)
+  @Test
   public void shouldGetPrimitiveArray() {
     SqlSession sqlSession = sqlSessionFactory.openSession();
     try {
-      // Throwing an exception is the expected behavior
-      // until #555 is fixed
       Mapper mapper = sqlSession.getMapper(Mapper.class);
       int[] ids = mapper.getUserIdsPrimitive();
       assertEquals(1, ids[0]);

